src/google/protobuf/compiler/cpp/cpp_message.cc

```diff
@@ -3080,7 +3080,10 @@ GenerateMergeFromCodedStream(io::Printer* printer) {
         } else {
           const FieldDescriptor* next_field = ordered_fields[i + 1];
           printer->Print(
-            "if (input->ExpectTag($next_tag$)) goto parse_$next_name$;\n",
+            "if (input->ExpectTag($next_tag$)) {\n"
+            "    tag = $next_tag$;\n"
```

Comment:
I think we can avoid this assignment by changing the enum field parsing code to not reference "tag". See:
https://github.com/google/protobuf/blob/master/src/google/protobuf/compiler/cpp/cpp_enum_field.cc#L145

Instead of emitting a variable name "tag", I think we can change that to emitting the actual tag int value. I.e.,
```cpp
printer->Print(
        "} else {\n"
        "  unknown_fields_stream.WriteVarint32($tag$);\n"
        "  unknown_fields_stream.WriteVarint32(value);\n",
        "tag", MakeTag(...));
```
Can you try that?